### PR TITLE
feat: EOF (ctrl-d) to shutdown the server gracefully

### PR DESCRIPTION
**Pull request by weihanglo**
Part of #62 

Support graceful shutdown the server by pressing Ctrl-d (EOF)

### CODE REVIEW

**Comment by weihanglo** [comment]
cc @sayanarijit
If all go well I'll release a new version.

**Comment by sayanarijit** [comment]
Can we log the `ctrl+d` instruction when the server loads?

**Comment by weihanglo** [comment]
There is a caveat I concern: is parent process pipe all stdin down the child process `sfz`, the vec would waste lots of space piling up of unnecessary data.

File: src/server/serve.rs [hunk 1]
```diff
@@ -68,8 +69,11 @@ pub async fn serve(args: Args) -> BoxResult<()> {
 
     let server = hyper::Server::try_bind(&address)?.serve(make_svc);
     let address = server.local_addr();
-    println!("Files served on http://{}{}", address, path_prefix);
-    server.await?;
+    println!(
```
Comment (sayanarijit) [inline src/server/serve.rs:72]: `eprintln` probably would be better.

**Comment by sayanarijit** [comment]
> There is a caveat I concern: if parent process pipes stdin down to the child process `sfz`, the vec would waste lots of space piling up of unnecessary data.

Sorry but can you pls clarify which `vec`?

File: src/server/serve.rs (hunk 1 above)
Comment (weihanglo) [reply src/server/serve.rs:72]: I am just thinking of this. Thank for this!

**Comment by weihanglo** [comment]
> > There is a caveat I concern: if parent process pipes stdin down to the child process `sfz`, the vec would waste lots of space piling up of unnecessary data.
> 
> Sorry but can you pls clarify which `vec`?

Here. Do you have any better approach?

https://github.com/weihanglo/sfz/pull/63/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR33

File: src/server/serve.rs (hunk 1 above)
Comment (weihanglo) [reply src/server/serve.rs]: fixed ef28b18

**Comment by sayanarijit** [comment]
> > > There is a caveat I concern: if parent process pipes stdin down to the child process `sfz`, the vec would waste lots of space piling up of unnecessary data.
> > 
> > 
> > Sorry but can you pls clarify which `vec`?
> 
> Here. Do you have any better approach?
> 
> https://github.com/weihanglo/sfz/pull/63/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR33

One way is to read bytes manually into a sized array until it returns 0.
Although. I'm sure why would anyone want to pipe data to a web server, unless intending to break things intentionally.

**Comment by weihanglo** [comment]
Got it. Seems trivial and we just skip it now.

**Comment by sayanarijit** [comment]
xplr-serve

Works fine.

**Comment by weihanglo** [comment]
Thanks @sayanarijit!

**Pull request merged**